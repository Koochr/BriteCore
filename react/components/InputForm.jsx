import React from "react"
import PropTypes from "prop-types"

export default class InputForm extends React.Component {
  static propTypes = {
  	initialValue: PropTypes.string,
  	onSubmit: PropTypes.func.isRequired,
  	onCancel: PropTypes.func,
  	submitButtonName: PropTypes.string,
  	cancelButtonName: PropTypes.string,
  }

  state = {
  	value: this.props.initialValue || "",
  }

  handleChange = event => {
  	this.setState({ value: event.target.value, })
  }

  handleSubmit = event => {
  	event.stopPropagation()
  	this.props.onSubmit(this.state.value)
  }

	handleCancel = event => {
		event.stopPropagation()
		this.props.onCancel()
	}

  onKeyPress = event => {
  	if (event.charCode === 13) {
  		event.preventDefault()
  		this.handleSubmit(event)
  	}
  }

  render() {
  	return <React.Fragment>
  		<form className="input-form">
  			<textarea
  				className="input-field form-control"
  				onChange={this.handleChange}
  				value={this.state.value}
  				onKeyPress={this.onKeyPress}
  			/>
  			<button
  				className="btn btn-success"
  				type="button"
  				onClick={this.handleSubmit}
  			>
  				{this.props.submitButtonName || "Submit" }
  			</button>
  			{this.props.onCancel && <button
  				className="btn btn-danger"
  				type="button"
  				onClick={this.handleCancel}>
  				{this.props.cancelButtonName || "Cancel" }
  			</button>}
  		</form>
  	</React.Fragment>
  }
}
