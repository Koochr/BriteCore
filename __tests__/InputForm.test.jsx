/* global it, jest, expect */
import React from "react"
import { shallow, configure, mount, } from "enzyme"
import Adapter from "enzyme-adapter-react-16"
import InputForm from "../react/components/InputForm.jsx"

configure({ adapter: new Adapter(), })

it("Should render correctly", () => {
	const component = shallow(
		<InputForm onSubmit={() => {}}/>
	)
	expect(component).toMatchSnapshot()
})

it("Should render needed elements", () => {
	const component = mount(
		<InputForm onSubmit={() => {}} onCancel={() => {}} submitButtonName="sbm-btn" cancelButtonName="cnc-btn" />
	)
	const form = component.find("form")
	expect(form).toHaveLength(1)
	const input = form.find("textarea")
	expect(input).toHaveLength(1)
	const buttonSubmit = component.find(".btn-success")
	expect(buttonSubmit).toHaveLength(1)
	expect(buttonSubmit.props().children).toStrictEqual("sbm-btn")
	const buttonCancel = component.find(".btn-danger")
	expect(buttonCancel).toHaveLength(1)
	expect(buttonCancel.props().children).toStrictEqual("cnc-btn")
})

it("Should handle input change and submit by button press", () => {
	const initialText = "some text"
	const component = mount(
		<InputForm onSubmit={jest.fn()} initialValue={initialText} />
	)
	const enteredText = " and some more text"
	const mockStopPropagation = jest.fn()
	expect(component.state().value).toStrictEqual(initialText)
	component.find("textarea").simulate("change", { target: { value: initialText + enteredText, }, })
	expect(component.state().value).toStrictEqual(initialText + enteredText)
	component.find("button").simulate("click", { stopPropagation: mockStopPropagation, })
	expect(component.props().onSubmit).toBeCalledWith(initialText + enteredText)
	expect(mockStopPropagation.mock.calls.length).toStrictEqual(1)
})

it("Should handle input change and submit by enter key press", () => {
	const initialText = "initial text"
	const component = mount(
		<InputForm onSubmit={jest.fn()} initialValue={initialText}/>
	)
	const enteredText = " and other text"
	const mockPreventDefault = jest.fn()
	const mockStopPropagation = jest.fn()
	expect(component.state().value).toStrictEqual(initialText)
	component.find("textarea").simulate("change", { target: { value: initialText + enteredText, }, })
	expect(component.state().value).toStrictEqual(initialText + enteredText)
	component.find("textarea").simulate("keyPress", {
		charCode: 13,
		preventDefault: mockPreventDefault,
		stopPropagation: mockStopPropagation,
	})
	expect(component.props().onSubmit).toBeCalledWith(initialText + enteredText)
	expect(mockPreventDefault.mock.calls.length).toStrictEqual(1)
	expect(mockStopPropagation.mock.calls.length).toStrictEqual(1)
})

it("Should handle cancel button press", () => {
	const component = mount(
		<InputForm onSubmit={jest.fn()} onCancel={jest.fn()} />
	)
	const mockStopPropagation = jest.fn()
	component.find(".btn-danger").simulate("click", { stopPropagation: mockStopPropagation, })
	expect(component.props().onCancel.mock.calls.length).toStrictEqual(1)
	expect(mockStopPropagation.mock.calls.length).toStrictEqual(1)

})
