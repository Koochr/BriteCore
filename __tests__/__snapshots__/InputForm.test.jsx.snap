// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InputForm
    onSubmit={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <form
        className="input-form"
      >
        <textarea
          className="input-field form-control"
          onChange={[Function]}
          onKeyPress={[Function]}
          value=""
        />
        <button
          className="btn btn-success"
          onClick={[Function]}
          type="button"
        >
          Submit
        </button>
      </form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <textarea
            className="input-field form-control"
            onChange={[Function]}
            onKeyPress={[Function]}
            value=""
          />,
          <button
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Submit
          </button>,
          undefined,
        ],
        "className": "input-form",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "input-field form-control",
            "onChange": [Function],
            "onKeyPress": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "className": "btn btn-success",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
        undefined,
      ],
      "type": "form",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <form
          className="input-form"
        >
          <textarea
            className="input-field form-control"
            onChange={[Function]}
            onKeyPress={[Function]}
            value=""
          />
          <button
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Submit
          </button>
        </form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <textarea
              className="input-field form-control"
              onChange={[Function]}
              onKeyPress={[Function]}
              value=""
            />,
            <button
              className="btn btn-success"
              onClick={[Function]}
              type="button"
            >
              Submit
            </button>,
            undefined,
          ],
          "className": "input-form",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "input-field form-control",
              "onChange": [Function],
              "onKeyPress": [Function],
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "btn btn-success",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
          undefined,
        ],
        "type": "form",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
